crates/trie/sparse/src/trie.rs

```diff
@@ -451,59 +416,48 @@ impl RevealedSparseTrie {
         masks: TrieMasks,
         retain_updates: bool,
     ) -> SparseTrieResult<Self> {
-        let mut this = Self {
-            nodes: HashMap::default(),
-            branch_node_tree_masks: HashMap::default(),
-            branch_node_hash_masks: HashMap::default(),
-            values: HashMap::default(),
-            prefix_set: PrefixSetMut::default(),
-            rlp_buf: Vec::new(),
-            updates: None,
-        }
-        .with_updates(retain_updates);
-        this.reveal_node(Nibbles::default(), root, masks)?;
-        Ok(this)
-    }
-}
-
-impl RevealedSparseTrie {
-    /// Sets the fields of this `RevealedSparseTrie` to the fields of the input
-    /// `SparseTrieState`.
-    ///
-    /// This is meant for reusing the allocated maps contained in the `SparseTrieState`.
-    ///
-    /// Copies over any existing nodes, branch masks, and values.
-    pub fn use_allocated_state(&mut self, mut other: SparseTrieState) {
-        for (path, node) in self.nodes.drain() {
-            other.nodes.insert(path, node);
-        }
-        for (path, mask) in self.branch_node_tree_masks.drain() {
-            other.branch_node_tree_masks.insert(path, mask);
-        }
-        for (path, mask) in self.branch_node_hash_masks.drain() {
-            other.branch_node_hash_masks.insert(path, mask);
-        }
-        for (path, value) in self.values.drain() {
-            other.values.insert(path, value);
-        }
-
-        self.nodes = other.nodes;
-        self.branch_node_tree_masks = other.branch_node_tree_masks;
-        self.branch_node_hash_masks = other.branch_node_hash_masks;
-        self.values = other.values;
+        Self::default().with_root(root, masks, retain_updates)
     }
 
     /// Configures the trie to retain information about updates.
     ///
     /// If `retain_updates` is true, the trie will record branch node updates and deletions.
     /// This information can then be used to efficiently update an external database.
     pub fn with_updates(mut self, retain_updates: bool) -> Self {
-        if retain_updates {
-            self.updates = Some(SparseTrieUpdates::default());
-        }
+        self.updates = retain_updates.then(|| {
+            self.updates
+                .map(|mut u| {
+                    u.clear();
+                    u
+                })
+                .unwrap_or_default()
+        });
```

Comment:
is this to reuse the updates vec?

Reply:
Yup, since this gets called as part of `with_root`, and the goal of all this is to reuse buffers, I thought it would be better to not throw away the updates buffer here.